loki/transformations/dependency.py

```diff
@@ -29,18 +29,135 @@ class DuplicateKernel(Transformation):
+            New item name.
+        local_name : str
+            New item local name.
+        config : :any:`SchedulerConfig`
+            The scheduler config to use when instantiating new items.
+        Returns
+        -------
+        :any:`Item`
+            Newly created item.
+        """
+        new_item = item_factory.item_cache.get(new_item_name)
+        # Try to get an item for the scope or create that first
+        if new_item is None and scope_name:
+            scope_item = item_factory.item_cache.get(scope_name)
+            if scope_item:
+                scope = scope_item.ir
+                if local_name not in scope and item.local_name in scope:
+                    # Rename the existing item to the new name
+                    scope[item.local_name].name = local_name
+
+                if local_name in scope:
+                    new_item = item_factory.create_from_ir(
+                        scope[local_name], scope, config=config
+                    )
+        # Create new item
+        if new_item is None:
+            new_item = item_factory.get_or_create_item_from_item(new_item_name, item, config=config)
+        return new_item
+
+    def _modify_sgraph(self, sgraph, item, new_items):
+        """
+        Add new items to graph.
+
+        Parameters
+        ----------
+        sgraph : :any:`SGraph`
+            Directed graph or rather copy of it to
+            be modified.
+        item : :any:`Item`
+            Node to which add the new items.
+        new_items : tuple
+            Tuple of :any:`Item` to add to graph.
+        """
+        sgraph.add_nodes(new_items)
+        sgraph.add_edges((item, _item) for _item in new_items)
+
+    def _rename_calls(self, new_item, new_dependencies):
+        """
+        Rename calls and imports according to the newly created
+        duplicated items.
+
+        Parameters
+        ----------
+        new_item : :any:`Item`
+            The newly created item for which calls and imports
+            to be renamed.
+        new_dependencies : dict
+            Dictionary used to get information about how
+            to rename calls and imports.
+        """
+        call_map = {}
+        for call in FindNodes(ir.CallStatement).visit(new_item.ir.body):
+            call_name = str(call.name).lower()
+            new_call_name = f'{call_name}{self.suffix}'.lower()
+            call_new_item = new_dependencies[new_call_name]
+            proc_symbol = call_new_item.ir.procedure_symbol.rescope(scope=new_item.ir)
+            call_map[call] = call.clone(name=proc_symbol)
+        # TODO: imports at module level ...
+        imp_map = {}
+        for imp in FindNodes(ir.Import).visit(new_item.ir.spec):
+            new_symbols = [symbol.clone(name=f'{symbol.name}{self.suffix}') for symbol in imp.symbols]
+            imp_map[imp] = imp.clone(module=f'{imp.module.lower()}{self.module_suffix}',
+                                     symbols=as_tuple(new_symbols))
```

Comment:
Does this really do the right thing? This looks like it blindly appends the suffix to every import and symbol, regardless of whether these are actually part of the replicated call tree (i.e., I think this would also rename a `use iso_fortran_env, only: real64` to `use iso_fortran_env{suffix}, only: real64{suffix}`). Maybe you can add a test for this?